connector/src/main/scala/com/datastax/spark/connector/datasource/CassandraCatalog.scala

```diff
@@ -0,0 +1,431 @@
+      case Failure(noSuchTableException: NoSuchTableException) => //We can create this table
+      case Failure(e) => throw e
+    }
+
+    //There is an implicit for this but it's only accessible in org.apache.spark.sql.catalog (maybe we should use it)
+    val invalidPartitions = partitions.filter(_.name() != "identity")
+    if (invalidPartitions.nonEmpty) {
+      throw new UnsupportedOperationException(s"Cassandra Tables can only by partitioned based on direct references to columns, found: ${invalidPartitions.mkString(",")}")
+    }
+
+    val providedPartitionKeyNames = partitions.map(_.references().head.fieldNames().head)
+
+    val partitionKeys = if (providedPartitionKeyNames.nonEmpty) {
+      providedPartitionKeyNames
+    } else {
+      tableProps
+        .getOrElse(PartitionKey, throw new CassandraCatalogException(s"Cassandra Tables need partition keys defined in property $PartitionKey or with 'PARTITIONED BY columns"))
+        .split(",")
+        .map(_.replaceAll("\\s", ""))
+    }
+
+    val partitionKeyNames = partitionKeys.map(fromInternal)
+
+    val clusteringKeyNames = tableProps
+      .get(ClusteringKey).toSeq
+      .flatMap(value => value.split(",").map(_.replaceAll("\\s", "").split("\\.")))
+      .map { case arr =>
+        if (arr.length != 1 && arr.length != 2)
+          throw new CassandraCatalogException(s"Unable to parse clustering column ${arr.mkString(".")}, too many components")
+        if (arr.length == 2) { //Ordering Passed
+          val clusteringOrder = Try(ClusteringOrder.valueOf(arr(1).toUpperCase())).getOrElse(throw new CassandraCatalogException(s"Invalid clustering order found in ${arr.mkString(".")}, must be ASC or DESC or blank"))
+          (fromInternal(arr(0)), clusteringOrder)
+        } else { //No Ordering Passed
+          (fromInternal(arr(0)), ClusteringOrder.ASC)
+        }
+      }
+
+    val protocolVersion = connector.withSessionDo(_.getContext.getProtocolVersion)
+
+    val columnToType = schema.fields.map(sparkField =>
+      (fromInternal(sparkField.name), sparkSqlToJavaDriverType(sparkField.dataType, protocolVersion))
+    ).toMap
+
+    val namespace = fromInternal(ident.namespace.head)
+    val table = fromInternal(ident.name())
+
+    val createTableStart: OngoingPartitionKey = SchemaBuilder.createTable(namespace, table)
+
+    val createTableWithPk: CreateTable = partitionKeyNames.foldLeft(createTableStart) { (createTable, pkName) =>
+      val dataType = columnToType.getOrElse(pkName,
+        throw new CassandraCatalogException(s"$pkName was defined as a partition key but it does not exist in the schema ${schema.fieldNames.mkString(",")}"))
+      createTable.withPartitionKey(pkName, dataType).asInstanceOf[OngoingPartitionKey]
+    }.asInstanceOf[CreateTable]
+
+    val createTableWithClustering = clusteringKeyNames.foldLeft(createTableWithPk) { (createTable, ckName) =>
+      val dataType = columnToType.getOrElse(ckName._1,
```

Comment:
I'd rather leave this till Spark gets a Sorted keyword. Since Spark is doing the heavy lifting for us at the moment by actually looking for the matching column identifier for Partitioned By ...

I'll add a little fix for now if that is ok